ComboboxControl: add unit tests

chad1008 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Add unit tests to the ComboboxControl component

## Why?
Following up on the conversation in #41417 where it was noted this component currently lacks any tests, making it harder to update it with confidence.

## How?
Five tests are currently implemented:
- test basic rendering of the component
- test that the label is visible when expected (the default)
- test that the label is hidden when expected (if `hideLabelFromVision` prop  is `true`)
- test that the correct list of options are rendered
- test that the correct option is selected when clicked

I did plan to add an additional test to validate the correct option is selected via keypresses, but ran into trouble implementing it. I wasn't able to get `userEffect.keyboard()` to register `downarrow` key presses, nor could I get it to simulate hitting `enter`. I was able to simulate typing in part of the option's name, but without `enter` to confirm the selection, the test wasn't viable.

I'm not 100% sure why that is, but I did find that [for some elements, certain keypresses (like arrowdown) aren't fully implemented](https://github.com/testing-library/user-event/issues/843#issuecomment-1068812316). This may be related?

## Testing Instructions
Run `npm run test-unit packages/components/src/combobox-control`

## Screenshots or screencast <!-- if applicable -->

chad1008 committed:
ComboboxControl: add initial rendering tests

chad1008 committed:
ComboboxControl: add test to validate rendered options

chad1008 committed:
ComboboxControl: add test to validate value after click

chad1008 committed:
ComboboxControl: use `getAllByRole` to identify rendered options

chad1008 commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,141 @@
+const getRenderedOptions = () => screen.getAllByRole( 'option' );
+
+describe( 'ComboboxControl', () => {
+	// Use real timers to avoid timeout errors from userEvent
+	beforeEach( () => {
+		jest.useRealTimers();
+	} );
```
This was needed because I found that `userEvent.click()` and `userEvent.keyboard()` would consistently time out when running the tests. This[ appears to be an issue with Jest 27.
](https://github.com/facebook/jest/issues/11607#issuecomment-899068995)

Because I wasn't able to fully implement what I'd planned with `userEvent.keyboard()` (see main issue description) we could fall back to `fireEvent` for this click. Since `userEvent()` was working as-is I left if, but I'm open to changing based on feedback!

ciampo replied:
We've been dealing with this problem in a different way:

1. To start, we should not call `userEvent` directly from the import — the [docs recommend](https://testing-library.com/docs/user-event/intro#writing-tests-with-userevent) using `userEvent.setup`
2. As part of that `userEvent.setup` call, we can pass `jest.advanceTimersByTime` as the function that `userEvent` should call to `advanceTimers` (this avoids the timeouts)

The resulting changes should look something like this:

```diff
diff --git a/packages/components/src/combobox-control/test/index.js b/packages/components/src/combobox-control/test/index.js
index f43c7422f2..e0bbac312d 100644
--- a/packages/components/src/combobox-control/test/index.js
+++ b/packages/components/src/combobox-control/test/index.js
@@ -58,12 +58,12 @@ const getLabel = () => screen.getByText( defaultLabelText );
 const getInput = () => screen.getByRole( 'combobox' );
 const getRenderedOptions = () => screen.getAllByRole( 'option' );
 
-describe( 'ComboboxControl', () => {
-	// Use real timers to avoid timeout errors from userEvent
-	beforeEach( () => {
-		jest.useRealTimers();
+const setupUser = () =>
+	userEvent.setup( {
+		advanceTimers: jest.advanceTimersByTime,
 	} );
 
+describe( 'ComboboxControl', () => {
 	const TestComboboxControl = ( props ) => {
 		const [ value, setValue ] = useState( null );
 		return (
@@ -124,11 +124,13 @@ describe( 'ComboboxControl', () => {
 	} );
 
 	it( 'should select the correct option with click', async () => {
+		const user = setupUser();
+
 		render( <TestComboboxControl /> );
 		const targetIndex = 2;
 		const input = getInput();
 		input.focus();
-		await userEvent.click(
+		await user.click(
 			screen.getByRole( 'option', {
 				name: timezones[ targetIndex ].name,
 			} )
```

The tests written in https://github.com/WordPress/gutenberg/blob/9a06585a7c125e8198feb299daabc8a507695c02/packages/components/src/select-control/test/select-control.tsx can be a good example to have a look at

chad1008 replied:
Thanks for this, I didn't know about the `advanceTimersByTime` trick!

Not sure how I missed that `userEvent.setup()` call. I _was_ using that on another test in this suite that didn't pan out, but somehow overlooked this implementation that was actually included in the PR. Thanks for catching it!

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,141 @@
+const timezoneOptions = timezones.map( mapTimezoneOption );
+const defaultLabelText = 'Select a timezone';
+const getLabel = () => screen.getByText( defaultLabelText );
+const getInput = () => screen.getByRole( 'combobox' );
```
Let's also check that the input is labelled correctly:

```suggestion
const getInput = ( name ) => screen.getByRole( 'combobox', { name } );
```

We can then call `getInput( defaultLabelText )` in our tests (or swap `defaultLabelText` with whatever label we need to test for)

chad1008 replied:
Good idea, thanks! 👍

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,141 @@
+const defaultLabelText = 'Select a timezone';
+const getLabel = () => screen.getByText( defaultLabelText );
+const getInput = () => screen.getByRole( 'combobox' );
+const getRenderedOptions = () => screen.getAllByRole( 'option' );
```
Nit, but I would rename this function to `getAllOptions`

chad1008 replied:
I didn't love `getRenderedOptions` so I'll happily implement this!

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,141 @@
+
+		expect( input.value ).toEqual( timezones[ targetIndex ].name );
+		expect( currentValue ).toEqual( timezones[ targetIndex ].abbreviation );
+	} );
```
I would write this test a bit differently:

- instead of parsing the value printed to the DOM in an element with `data-testid`, I would use pass a spy to the `onChange` prop, and observe the value passed to it
- instead of using the imperative `focus()` method on the `input` element, I would focus it by simulating a user's click. This is much closer to how the end user would use this component than firing a synthetic focus event
- (nit) I would refactor out a utility function to select an `option` element, given a specific name
- I would add inline comments to explain what is going on in the test
- (another nit) I would use the `toHaveValue` assertion, instead of manually checking if `input.value` is equal to the expected value.


Here's how the code could look like after applying the notes left above:

<details>

<summary>Click to expand</summary>

```diff
diff --git a/packages/components/src/combobox-control/test/index.js b/packages/components/src/combobox-control/test/index.js
index f43c7422f2..ed28793d4e 100644
--- a/packages/components/src/combobox-control/test/index.js
+++ b/packages/components/src/combobox-control/test/index.js
@@ -55,7 +55,8 @@ const mapTimezoneOption = ( tz ) => ( {
 const timezoneOptions = timezones.map( mapTimezoneOption );
 const defaultLabelText = 'Select a timezone';
 const getLabel = () => screen.getByText( defaultLabelText );
-const getInput = () => screen.getByRole( 'combobox' );
+const getInput = () => screen.getByRole( 'combobox', { name: defaultLabelText } );
+const getOption = ( name ) => screen.getByRole( 'option', { name } );
 const getRenderedOptions = () => screen.getAllByRole( 'option' );
 
 describe( 'ComboboxControl', () => {
@@ -64,8 +65,13 @@ describe( 'ComboboxControl', () => {
 		jest.useRealTimers();
 	} );
 
-	const TestComboboxControl = ( props ) => {
+	const TestComboboxControl = ( { onChange, ...props } ) => {
 		const [ value, setValue ] = useState( null );
+		const handleOnChange = ( value ) => {
+			setValue( value );
+			onChange?.( value );
+		}
+
 		return (
 			<>
 				<ComboboxControl
@@ -73,9 +79,8 @@ describe( 'ComboboxControl', () => {
 					value={ value }
 					label={ defaultLabelText }
 					options={ timezoneOptions }
-					onChange={ setValue }
+					onChange={ handleOnChange }
 				/>
-				<p data-testid="value-output">{ value }</p>
 			</>
 		);
 	};
@@ -123,19 +128,22 @@ describe( 'ComboboxControl', () => {
 		expect( renderedOptionNames ).toEqual( timezoneNames );
 	} );
 
-	it( 'should select the correct option with click', async () => {
-		render( <TestComboboxControl /> );
-		const targetIndex = 2;
+	it( 'should select the correct option via click events', async () => {
+		const targetOption = timezones[ 2 ];
+		const onChangeSpy = jest.fn();
+
+		render( <TestComboboxControl onChange={ onChangeSpy } /> );
+
 		const input = getInput();
-		input.focus();
-		await userEvent.click(
-			screen.getByRole( 'option', {
-				name: timezones[ targetIndex ].name,
-			} )
-		);
-		const currentValue = screen.getByTestId( 'value-output' ).textContent;
 
-		expect( input.value ).toEqual( timezones[ targetIndex ].name );
-		expect( currentValue ).toEqual( timezones[ targetIndex ].abbreviation );
+		// Clicking on the input shows the options
+		await userEvent.click( input );
+
+		// Select the target option
+		await userEvent.click( getOption( targetOption.name ) );
+
+		expect( onChangeSpy ).toHaveBeenCalledTimes( 1 );
+		expect( onChangeSpy ).toHaveBeenCalledWith( targetOption.abbreviation );
+		expect( input ).toHaveValue( targetOption.name );
 	} );
 } );
```
</details>

chad1008 replied:
🤯 This is awesome, thank you for laying it all out. I think I just learned about 6 new things.

I've implemented the above with a small tweak... when defining `handleOnchange` I had to use `newValue` for the param name, because `value` was already declared in the upper scope, triggering a lint error.

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,141 @@
+		render( <TestComboboxControl /> );
+
+		const input = getInput();
+		expect( input ).toBeVisible();
```
Interesting. My default assertion in this case is `toBeInTheDocument()`, but I also see why `toBeVisible` makes sense. Maybe we should add both, to make it more explicit when we later check for the label to be in the document, but not visible?

WDYT, @mirka ?

```suggestion
		expect( input ).toBeInTheDocument();
		expect( input ).toBeVisible();
```

chad1008 replied:
This makes sense to me, but I'll wait for @mirka's input as well 😄

mirka replied:
No strong opinion on the specifics, but I generally want the `it()` description to match what is actually being tested. "It should render" to me isn't a particularly useful description in this context. Should it say "It should render the input in the document" or "It should render a visible input in the document"? (But if that is actually what we do mean to test, I would find it more economical if that were tested as part of a longer interaction test, e.g. typing in the input and filtering options.)

chad1008 replied:
Thank you both... I've incorporated both bits of feedback now.

- I've gone ahead and removed the initial `should render` test, which was a pattern I was borrowing from another suite. I do agree that it's more economical to test that as part of other tests, which happens naturally anyway though!
- With that test removed, I did implement Marco's suggestion for presence of the label, and then its visibility.

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,141 @@
+
+	it( 'should render with visible label', () => {
+		render( <TestComboboxControl /> );
+		const label = getLabel();
+		expect( label ).toBeVisible();
```
Seeing this test, it makes me think that by default the `<TestComboboxControl />` should NOT add a `label` to `ComboboxControl` — it should be possible to pass `label` as a prop to `TestComboboxControl`

chad1008 replied:
Cool. Along with this change, I also updated `getLabel()` to accept any label text, since `defaultLabelText` won't be the automatic default any longer. This way we can pass any label we need for a test, and then make sure it's rendering that text correctly.

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,141 @@
+		const label = getLabel();
+
+		expect( label ).toBeInTheDocument();
+		expect( label ).toHaveAttribute(
+			'data-wp-component',
+			'VisuallyHidden'
+		);
```
This code is testing implementation details — what we really want to test, is that the label can't be seen by the user. Did you try to 

```suggestion
		expect( label ).not.toBeVisible();
```

?

chad1008 replied:
I did! Unfortunately, `not.toBeVisible()` doesn't work here, because we're not doing [any of the things that satisfy that assertion](https://github.com/testing-library/jest-dom#tobevisible). Instead, our method of hiding the label relies on some clipping and absolute positioning, but as far as the test is concerned it's still technically visible.

Is that something we should look into addressing in a followup, so we can use a less implementation-focused test?

mirka replied:
Wow, I hadn't considered that! I found at least two existing tests that have hit this problem ([BorderControl](https://github.com/WordPress/gutenberg/blob/aa221c7ced2b68b019b771bb2f0c90627709833b/packages/components/src/border-control/test/index.js#L105-L108), [BorderBoxControl](https://github.com/WordPress/gutenberg/blob/aa221c7ced2b68b019b771bb2f0c90627709833b/packages/components/src/border-box-control/test/index.js#L103-L106)).

I can't really think of an easy way to address it though 🤔 Maybe it is a minor inconvenience we have to live with.

ciampo replied:
I guess so. Given that we have [a snapshot test](https://github.com/WordPress/gutenberg/blob/d5de22f316c1049f46675b52e1436890a16d7a5d/packages/components/src/visually-hidden/test/index.tsx) for `VIsuallyHidden` printing out the exact styles, I guess we can leave the code in this test as-is for the time being

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,141 @@
+	it( 'should render with the correct options', () => {
+		render( <TestComboboxControl /> );
+
+		getInput().focus();
```
Same as previous comment — let's prefer an explicit `user.click()`, to an imperative, synthetic focus event.

chad1008 replied:
👍

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,141 @@
+		}
+		const timezoneNames = timezones.map( ( tz ) => tz.name );
+
+		expect( renderedOptionNames ).toEqual( timezoneNames );
```
If I understand what you're trying to achieve here, it could be written in a simpler way:

```suggestion
		// Confirm the rendered options match the provided dataset.
		expect( renderedOptions ).toHaveLength( timezones.length );
		renderedOptions.forEach( ( option, optionIndex) => {
			expect( option ).toHaveTextContent( timezones[optionIndex].name )
		} );
```

chad1008 replied:
Much cleaner, thank you!

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,141 @@
+ */
+import ComboboxControl from '../';
+
+const timezones = [
+	{ name: 'Greenwich Mean Time', abbreviation: 'GMT' },
+	{ name: 'Universal Coordinated Time', abbreviation: 'UTC' },
+	{ name: 'European Central Time', abbreviation: 'ECT' },
+	{ name: '(Arabic) Egypt Standard Time', abbreviation: 'ART' },
+	{ name: 'Eastern African Time', abbreviation: 'EAT' },
+	{ name: 'Middle East Time', abbreviation: 'MET' },
+	{ name: 'Near East Time', abbreviation: 'NET' },
+	{ name: 'Pakistan Lahore Time', abbreviation: 'PLT' },
+	{ name: 'India Standard Time', abbreviation: 'IST' },
+	{ name: 'Bangladesh Standard Time', abbreviation: 'BST' },
+	{ name: 'Vietnam Standard Time', abbreviation: 'VST' },
+	{ name: 'China Taiwan Time', abbreviation: 'CTT' },
+	{ name: 'Japan Standard Time', abbreviation: 'JST' },
+	{ name: 'Australia Central Time', abbreviation: 'ACT' },
+	{ name: 'Australia Eastern Time', abbreviation: 'AET' },
+	{ name: 'Solomon Standard Time', abbreviation: 'SST' },
+	{ name: 'New Zealand Standard Time', abbreviation: 'NST' },
+	{ name: 'Midway Islands Time', abbreviation: 'MIT' },
+	{ name: 'Hawaii Standard Time', abbreviation: 'HST' },
+	{ name: 'Alaska Standard Time', abbreviation: 'AST' },
+	{ name: 'Pacific Standard Time', abbreviation: 'PST' },
+	{ name: 'Phoenix Standard Time', abbreviation: 'PNT' },
+	{ name: 'Mountain Standard Time', abbreviation: 'MST' },
+	{ name: 'Central Standard Time', abbreviation: 'CST' },
+	{ name: 'Eastern Standard Time', abbreviation: 'EST' },
+	{ name: 'Indiana Eastern Standard Time', abbreviation: 'IET' },
+	{ name: 'Puerto Rico and US Virgin Islands Time', abbreviation: 'PRT' },
+	{ name: 'Canada Newfoundland Time', abbreviation: 'CNT' },
+	{ name: 'Argentina Standard Time', abbreviation: 'AGT' },
+	{ name: 'Brazil Eastern Time', abbreviation: 'BET' },
+	{ name: 'Central African Time', abbreviation: 'CAT' },
+];
+const mapTimezoneOption = ( tz ) => ( {
+	value: tz.abbreviation,
+	label: tz.name,
+} );
```
Is there a reason why we don't just use `value` and `label` in the `timezones` variable, instead of mapping it later on?

chad1008 replied:
You make a good point! There isn't a real reason, I initially started this PR by following the pattern of [the example dataset in Storybook](https://github.com/WordPress/gutenberg/blob/8011374a3b7f17e21c7515375bc775e50d2e6888/packages/components/src/combobox-control/stories/index.js#L262)... but now that you point it out I can definitely see how I'm just adding an unnecessary `map` call. Thanks!

ciampo reviewed (commented):
Thank you @chad1008 for working on this! Adding unit tests to this component will be of great value.

Regarding what other tests we could add, here's what comes to mind:
- we should test both the controlled and uncontrolled version of the component
- we should add tests in which the user interacts with the component exclusively using the keyboard
- We should add a test for a more complete and typical user interaction, involving search and filtering. For example:
  - user clicks on the combobox
  - (assert that all options are being rendered)
  - user moves focus on the search input
  - user types a string that should not generate any matches
  - (assert that there are no options being rendered)
  - user clears text search input
  - (assert that all options are being rendered)
  - user types a new search string, this time one that should match a few options
  - (assert that the options that are being rendered are the expected ones matching the search string)
  - user clicks on one of the available options
  - (assert that input value and onChange spy have the new correct value selected)
  - etc...


A few important reflections that I noticed while reviewing this component:

1. The last test has revealed a potential discrepancy: it looks like the `<input />`'s `value` and the argument passed to the `onChange` callback are not the same value 😱  Shouldn't the expectation be that the `input`'s `value` is the `abbreviation`, instead of the `name`?between the value passed to `onChange` and the value of the `input
2. Not sure if this is an accessibility issue, but while reviewing the tests, I noticed that the `options` are not rendered to the DOM until the combobox is focused/clicked. This differs from other `Combobox` examples where those option items are rendered in the DOM, but just not visible (e.g. [`ariakit`](https://ariakit.org/examples/combobox))
3. `userEvent` has some nice utility functions, including one to [select or deselect `option`s](https://testing-library.com/docs/user-event/utility#-selectoptions-deselectoptions) — unfortunately I tried it for these tests, but it doesn't seem to work because we're using custom aria roles, instead of native `select` and `option` elements.

cc @mirka , especially regarding the last 3 points.

We should also keep in mind that this component is potentially going to be rewritten using ariakit component primitives.

ciampo commented:
> we should test both the controlled and uncontrolled version of the component

@stokesman , if I remember correctly you had worked a few weeks ago on adding some unit tests for a component in which you were testing for both the controlled and uncontrolled version of such component?

stokesman commented:
>testing for both the controlled and uncontrolled version of such component?

Yes, it was `RangeControl` and the way I'd approached was turning the outer `describe` into [`describe.each`](https://jestjs.io/docs/api#describeeachtablename-fn-timeout). The change didn't end up in trunk but here’s the commit https://github.com/WordPress/gutenberg/pull/40568/commits/ce0ac420f22da5afc8b750e12e850f7e4dc49110.

chad1008 committed:
ComboboxContorl: implement `userEvent.setup()` and `jest.advanceTimersbyTime`

chad1008 committed:
ComboboxControl: update dataset object keys

chad1008 committed:
ComboboxControl: ensure unit tests validate the input label

chad1008 committed:
ComboboxControl: rename `getRenderedOptions` function

chad1008 committed:
ComboboxControl: add `onChange` prop to unit tests, refactor click event selection test

chad1008 committed:
ComboboxControl: unit tests - `label` prop is no longer passed by default

chad1008 committed:
ComboboxControl: unit tests - replace `focus()` with `userEvent.click()`

chad1008 committed:
ComboboxControl: refactor test assertion

chad1008 commented:
Wow, thank you for the thorough review at @ciampo... really informative and super helpful.

I've pushed some updates addressing most of your points. Still to do:

- [x] we should test both the controlled and uncontrolled version of the component
- [x] we should add tests in which the user interacts with the component exclusively using the keyboard
- [x] we should add a test for a more complete and typical user interaction, involving search and filtering

The second item in that list will be tricky. Thus far I haven't had any luck getting `userEvent.keyboard()` to successfully register arrow keys or the `enter` keys for this component. Typing works fine, but navigating and and actually selecting an option remains elusive.

I'm not sure why yet, but[ I have seen reports of issues with arrow keys on other elements](https://github.com/testing-library/user-event/issues/843#issuecomment-1068812316), which could be related (or possibly not?)

chad1008 commented:
> Yes, it was `RangeControl` and the way I'd approached was turning the outer `describe` into [`describe.each`](https://jestjs.io/docs/api#describeeachtablename-fn-timeout). The change didn't end up in trunk but here’s the commit [ce0ac42](https://github.com/WordPress/gutenberg/commit/ce0ac420f22da5afc8b750e12e850f7e4dc49110).

Nice, thank you @stokesman!

mirka reviewed (commented):
> 2\. Not sure if this is an accessibility issue, but while reviewing the tests, I noticed that the `options` are not rendered to the DOM until the combobox is focused/clicked. This differs from other `Combobox` examples where those option items are rendered in the DOM, but just not visible (e.g. [`ariakit`](https://ariakit.org/examples/combobox))

I think it should be fine, given that options can often be lazy loaded (e.g. virtualized or paginated).

chad1008 committed:
ComboboxControl: add keypress unit test

chad1008 committed:
ComboboxControl: unit test - fix misnamed variable

chad1008 committed:
ComboboxControl: add search unit test

chad1008 commented:
Interesting development I'd love some input on. I was writing a test for invalid search strings and found the component currently behaves differently than I'd expect.

When the input gets focus, it starts by highlighting the zero indexed element of the list, which makes sense. As the user types it filters that list. If a string with no matches is entered, I expected it would set the value to an empty string (selecting nothing).

It actually selects the original zero-indexed element (in the case of these tests, Greenwich Mean Time). So you can type something that doesn't exist, hit `Enter` and get a value that doesn't even closely resemble what you searched for.

That feels like a bug to address in a separate PR to me... thoughts @ciampo / @mirka ?

_Edit/correction:_ On invalid strings, it actually selects whatever the last highlighted item was (not just the zero index), either because of a hover or a partially correct string. For example, in Storybook, typing `Algzzzzz` and hitting `Enter` would actually select `Algeria`, because that was the last valid match before the string started failing.

chad1008 commented:
I dug a little more and this [appears to be intentional](https://github.com/WordPress/gutenberg/blob/10f8542e26e83262f177cdcca30ca57ae516e644/packages/components/src/combobox-control/index.js#L186-L189). The list is filtered as the user types and invalid matches choose the zero index of the remaining option(s). The goal of [this PR](https://github.com/WordPress/gutenberg/pull/33928) appears to be to address confusion when an existing option was already selected a new string was entered.

This still feels like a confusing experience to me though, because when I type an invalid string, the suggestion I'm ultimately selecting isn't shown until after I select it.

I can think of three options:
1. Change the behavior so that matchless searches return empty (bonus, we could look into replacing the suggestion list with a "No matches found" indicator)
2. Keep the behavior as-is but show the 'fallback suggestion' that will ultimately get selected for non-matching strings
3. No changes, invalid searches continue to choose the last remaining match without any suggestions shown

ciampo commented:
> I dug a little more and this [appears to be intentional](https://github.com/WordPress/gutenberg/blob/10f8542e26e83262f177cdcca30ca57ae516e644/packages/components/src/combobox-control/index.js#L186-L189). The list is filtered as the user types and invalid matches choose the zero index of the remaining option(s). The goal of [this PR](https://github.com/WordPress/gutenberg/pull/33928) appears to be to address confusion when an existing option was already selected a new string was entered.
> 
> This still feels like a confusing experience to me though, because when I type an invalid string, the suggestion I'm ultimately selecting isn't shown until after I select it.
> 
> I can think of three options:
> 
> 1. Change the behavior so that matchless searches return empty (bonus, we could look into replacing the suggestion list with a "No matches found" indicator)
> 2. Keep the behavior as-is but show the 'fallback suggestion' that will ultimately get selected for non-matching strings
> 3. No changes, invalid searches continue to choose the last remaining match without any suggestions shown

This definitely looks like an aspect of the component that we can polish — but I would do it in a separate issue. Let's keep this issue focused on adding the essential unit tests to `ComboboxControl` as planned.

> The last test has revealed a potential discrepancy: it looks like the <input />'s value and the argument passed to the onChange callback are not the same value 😱 Shouldn't the expectation be that the input's value is the abbreviation, instead of the name?between the value passed to onChange and the value of the `input

Just flagging this point in case it got overlooked — this may be an even higher priority task to follow-up with after this PR (and the related exhaustive-deps changes) get merged.

chad1008 committed:
remove extraneous test

chad1008 committed:
remove outdated assertion statement

chad1008 committed:
add multiple interaction test

chad1008 committed:
abstract setTargetOption helper function

chad1008 committed:
fix outdated variable names

chad1008 committed:
add controlled vs uncontrolled testing

chad1008 committed:
add selection announcement test

chad1008 committed:
assert that visible label should be in the document

chad1008 commented:
I've just pushed some updates to address the feedback shared above! I responded to one of the comments directly, but in addition to that I've added additional tests for
- searching/filtering
- complex interactions
- keyboard only interaction
- controlled vs. uncontrolled modes (Kudos to @stokesman for inspiration on the `describe.each` approach 👏 )
- confirming aria announcements are rendered upon selection 

Important note: some of the uncontrolled tests **will fail** until https://github.com/WordPress/gutenberg/pull/42752 is merged, as the component's current implementation doesn't support an uncontrolled mode. Without a value prop passed in, the component never actually inserts a selection into the input element. The linked PR addresses that (hat tip to @ciampo for teaching me how to go about implementing that hook!)

ciampo commented:
Sounds good! Let's merge #42752 first, rebase this PR, and then give it a final round of review.

chad1008 commented:
https://github.com/WordPress/gutenberg/pull/42752 merge and a rebase of this PR are complete. Marking as ready for review!

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,311 @@
+const getInput = ( name ) => screen.getByRole( 'combobox', { name } );
+const getOption = ( name ) => screen.getByRole( 'option', { name } );
+const getAllOptions = () => screen.getAllByRole( 'option' );
+const setTargetOption = ( index ) => {
```
This comment is a mix of nitpicking and general advice:

- I believe a more appropriate name for this function would be `getTargetOption`, as you're passing an index and getting the option in return
- Since we're setting the `options` prop in each test, it feels more correct that this function takes the array of options as a parameter too, instead of assuming that to always be `timezones`
- Overall, I'm not sure about how much value does this abstraction holds. It basically exists in order to append the `searchString` property. At most, if we wanted to abstract that piece of functionality away, we could at most create a utility function like `getSearchString( option ) => option.label.substring( 0, 11 )` — or even have no "utility" function at all

chad1008 replied:
Agreed on the function name and the `options` parameter. Made both of those changes! 👍 

Re: the abstraction itself... in addition to appending the `searchString` property, I also felt like it could be beneficial from a maintenance perspective to have the desired index declared once per test.

The target property gets referenced as much as three times per test. Without the abstraction, if we ever update the target of a given test we need to make sure to hit all three instances of it. With the abstraction it can just be updated once for the test in question, because the abstraction keeps it nice and DRY.

I'm happy to remove it or replace it with a `getSearchString()` if you think that would be best though.

ciampo replied:
> Re: the abstraction itself... in addition to appending the searchString property, I also felt like it could be beneficial from a maintenance perspective to have the desired index declared once per test.
>
> The target property gets referenced as much as three times per test. Without the abstraction, if we ever update the target of a given test we need to make sure to hit all three instances of it. With the abstraction it can just be updated once for the test in question, because the abstraction keeps it nice and DRY.

For completeness sake, here's what I had in mind when writing the previous message:

<details>

<summary>click to expand</summary>

```diff
diff --git a/packages/components/src/combobox-control/test/index.js b/packages/components/src/combobox-control/test/index.js
index 2bcd980982..72033f679f 100644
--- a/packages/components/src/combobox-control/test/index.js
+++ b/packages/components/src/combobox-control/test/index.js
@@ -53,10 +53,7 @@ const getLabel = ( labelText ) => screen.getByText( labelText );
 const getInput = ( name ) => screen.getByRole( 'combobox', { name } );
 const getOption = ( name ) => screen.getByRole( 'option', { name } );
 const getAllOptions = () => screen.getAllByRole( 'option' );
-const getTargetOption = ( options, index ) => {
-	const searchString = options[ index ].label.substring( 0, 11 );
-	return { searchString, ...options[ index ] };
-};
+const getOptionSearchString = ( option ) => option.label.substring( 0, 11 );
 const setupUser = () =>
 	userEvent.setup( {
 		advanceTimers: jest.advanceTimersByTime,
@@ -138,7 +135,7 @@ describe.each( [
 
 	it( 'should select the correct option via click events', async () => {
 		const user = setupUser();
-		const targetOption = getTargetOption( timezones, 2 );
+		const targetOption = timezones[ 2 ];
 		const onChangeSpy = jest.fn();
 		render(
 			<Component
@@ -163,7 +160,7 @@ describe.each( [
 	it( 'should select the correct option via keypress events', async () => {
 		const user = setupUser();
 		const targetIndex = 4;
-		const targetOption = getTargetOption( timezones, targetIndex );
+		const targetOption = timezones[ targetIndex ];
 		const onChangeSpy = jest.fn();
 		render(
 			<Component
@@ -192,7 +189,7 @@ describe.each( [
 
 	it( 'should select the correct option from a search', async () => {
 		const user = setupUser();
-		const targetOption = getTargetOption( timezones, 13 );
+		const targetOption = timezones[ 13 ];
 		const onChangeSpy = jest.fn();
 		render(
 			<Component
@@ -207,7 +204,7 @@ describe.each( [
 		await user.tab();
 
 		// Type enough characters to ensure a predictable search result
-		await user.keyboard( targetOption.searchString );
+		await user.keyboard( getOptionSearchString( targetOption ) );
 
 		// Pressing Enter/Return selects the currently focused option
 		await user.keyboard( '{Enter}' );
@@ -219,7 +216,7 @@ describe.each( [
 
 	it( 'should render aria-live announcement upon selection', async () => {
 		const user = setupUser();
-		const targetOption = getTargetOption( timezones, 9 );
+		const targetOption = timezones[ 9 ];
 		const onChangeSpy = jest.fn();
 		render(
 			<Component
@@ -233,7 +230,7 @@ describe.each( [
 		await user.tab();
 
 		// Type enough characters to ensure a predictable search result
-		await user.keyboard( targetOption.searchString );
+		await user.keyboard( getOptionSearchString( targetOption ) );
 
 		// Pressing Enter/Return selects the currently focused option
 		await user.keyboard( '{Enter}' );
@@ -248,7 +245,7 @@ describe.each( [
 	it( 'should process multiple entries in a single session', async () => {
 		const user = setupUser();
 		const unmatchedString = 'Mordor';
-		const targetOption = getTargetOption( timezones, 6 );
+		const targetOption = timezones[ 6 ];
 		const onChangeSpy = jest.fn();
 		render(
 			<Component
@@ -289,12 +286,13 @@ describe.each( [
 		} );
 
 		// Run a second search with a valid string.
-		await user.keyboard( targetOption.searchString );
+		const searchString = getOptionSearchString( targetOption );
+		await user.keyboard( searchString );
 		const validSearchRenderedOptions = getAllOptions();
 
 		// Find option that match the search string.
 		const matches = timezones.filter( ( option ) =>
-			option.label.includes( targetOption.searchString )
+			option.label.includes( searchString )
 		);
 
 		// Confirm the rendered options match the provided dataset based on the current string.
```

</details>

Removing the `getTargetOption` abstraction doesn't necessarily mean that the `targetOption` can be saved as a variable in each test and invoked multiple times. We could then call the `getOptionSearchString` only where it's actually necessary to get a search string.

Anyway, not a big deal — we can merge this PR with or without the suggested changes

chad1008 replied:
Ah, brilliant. My concerns about repetition were more based on your suggestion of possible having no utility function at all. I really do like your proposed `getOptionSearchString()` approach, it's a lot cleaner than my previous implementation.

Adding your changes to the PR now, which should be the final piece of the puzzle for this one. Thank you @ciampo!

ciampo commented on packages/components/src/combobox-control/test/index.js:
```diff
@@ -0,0 +1,311 @@
+		// Pressing Enter/Return selects the currently focused option
+		await user.keyboard( '{Enter}' );
+
+		const itemSelectedAnnouncement = screen.getByText( 'Item selected.' );
+
+		expect( itemSelectedAnnouncement ).toBeInTheDocument();
```
We're not really testing that an `aria-live` region is being rendered. Since `testing-library` doesn't quite support querying by attribute or specifically querying for `aria-live` regions, the best we can do is use the `selector` option to specify that the element that we're looking for should have the `aria-live` attribute set.

```suggestion
		expect(
			screen.getByText( 'Item selected.', {
				selector: '[aria-live]',
			} )
		).toBeInTheDocument();
```

chad1008 replied:
This is great, I didn't love settling for just the text content. Adding the `selector` option to my bag of tricks, thank you!

ciampo reviewed (commented):
Looking good! I just left a couple more comments.

Could you also add a CHANGELOG entry? This could go under the `Internal` section, I believe

chad1008 committed:
update CHANGELOG

chad1008 committed:
rename setTargetOption to getTargetOption, add options param

chad1008 committed:
Update packages/components/src/combobox-control/test/index.js

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

ciampo reviewed (approved):
LGTM 🚀 

(there's still one pending [comment](https://github.com/WordPress/gutenberg/pull/42403#discussion_r940160883), but it's not a blocker and we can merge this PR with or without the changes suggested in there)

chad1008 committed:
ComboboxControl: replace unit test util `getTargetOption` with `getOptionSearchString`

chad1008 commented:
Thanks @ciampo. The changes recommended in that comment are now included in the PR 🙂. Once tests pass I'll get this one merged. Thanks again for all of the feedback!
